docs/rfcs/GPUXDialect.md

```diff
@@ -0,0 +1,226 @@
+  call @fillResource1DFloat(%2, %cst_1) : (memref<?xf32>, f32) -> ()
+  gpu.launch_func  @main_kernel::@main_kernel blocks in (%c8, %c1, %c1) threads in (%c1, %c1, %c1) args(%0 : memref<8xf32>, %1 : memref<8xf32>, %2 : memref<8xf32>)
+  %3 = memref.cast %2 : memref<8xf32> to memref<*xf32>
+  call @printMemrefF32(%3) : (memref<*xf32>) -> ()
+  gpu.dealloc(%0) : memref<8xf32>
+  gpu.dealloc(%1) : memref<8xf32>
+  gpu.dealloc(%2) : memref<8xf32>
+  return
+}
+
+```
+
+// IR after our custom transformation pass:
+
+```
+
+func.func @main(%stream : !gpux.StreamType) attributes {llvm.emit_c_interface} {
+  %c8 = arith.constant 8 : index
+  %c1 = arith.constant 1 : index
+  %cst = arith.constant 2.200000e+00 : f32
+  %cst_0 = arith.constant 1.100000e+00 : f32
+  %cst_1 = arith.constant 0.000000e+00 : f32
+  %0 = gpux.alloc (%stream) : (!gpux.StreamType) -> memref<8xf32>
+  %1 = gpux.alloc (%stream) : (!gpux.StreamType) -> memref<8xf32>
+  %2 = gpux.alloc  (%stream) : (!gpux.StreamType) -> memref<8xf32>
+  call @fillResource1DFloat(%0, %cst_0) : (memref<?xf32>, f32) -> ()
+  call @fillResource1DFloat(%1, %cst) : (memref<?xf32>, f32) -> ()
+  call @fillResource1DFloat(%2, %cst_1) : (memref<?xf32>, f32) -> ()
+  gpux.launch_func  @main_kernel::@main_kernel blocks in (%c8, %c1, %c1) threads in (%c1, %c1, %c1) args(%stream : !gpux.StreamType, %memref : memref<8xf32>, %memref_2 : memref<8xf32>, %memref_3 : memref<8xf32>)
+  %3 = memref.cast %memref_3 : memref<8xf32> to memref<*xf32>
+  call @printMemrefF32(%3) : (memref<*xf32>) -> ()
+  gpux.dealloc(%stream, %0) : (!gpux.StreamType, memref<8xf32>) -> ()
+  gpux.dealloc(%stream, %1) : (!gpux.StreamType, memref<8xf32>) -> ()
+  gpux.dealloc(%stream, %2) : (!gpux.StreamType, memref<8xf32>) -> ()
+  return
+}
+```
+
+Since the stream is already provided by the user in this case, only the alloc, dealloc and the launch_func operations were translated to the extended versions in the GPUX dialect.
+
+
+## Example Usage in case of default stream:
+
+// IR before our custom transformation pass:
+
+```
+
+func.func @main() attributes {llvm.emit_c_interface} {
+  %c8 = arith.constant 8 : index
+  %c1 = arith.constant 1 : index
+  %cst = arith.constant 2.200000e+00 : f32
+  %cst_0 = arith.constant 1.100000e+00 : f32
+  %cst_1 = arith.constant 0.000000e+00 : f32
+  %0 = gpu.alloc  () : memref<8xf32>
+  %1 = gpu.alloc  () : memref<8xf32>
+  %2 = gpu.alloc  () : memref<8xf32>
+  gpu.launch_func  @main_kernel::@main_kernel blocks in (%c8, %c1, %c1) threads in (%c1, %c1, %c1) args(%0 : memref<8xf32>, %1 : memref<8xf32>, %2 : memref<8xf32>)
+  %3 = memref.cast %2 : memref<8xf32> to memref<*xf32>
+  call @printMemrefF32(%3) : (memref<*xf32>) -> ()
+  gpu.dealloc(%0) : memref<8xf32>
+  gpu.dealloc(%1) : memref<8xf32>
+  gpu.dealloc(%2) : memref<8xf32>
+  return
+}
+
+```
+
+// IR after our custom transformation pass:
+
+```
+
+func.func @main() attributes {llvm.emit_c_interface} {
```

Comment:
This pass makes sense to me.  This means user doesn't want to control at all.

Reply:
Why not? From our perspective that's exactly what happens with compute-follows-data. There can be a default behavior, but it seems important to be able to control this.